File: code-tiny/tools/graph/schema/preflight.py
"""Automatic, fail-closed graph schema preflight."""

from __future__ import annotations

import asyncio
import logging
import time
from dataclasses import dataclass
from typing import Any, Dict, Iterable, Mapping, Optional, Sequence, Tuple

from .manifest import CODE_GRAPH_SCHEMA, GraphSchemaManifest, SchemaIndex


logger = logging.getLogger(__name__)

_READY_STATES = {"ONLINE", "OPERATIONAL", "READY"}
_FAILED_STATES = {"FAILED", "FAILURE", "ERROR"}


class SchemaPreflightError(RuntimeError):
    """Required graph schema could not be made operational."""


@dataclass(frozen=True)
class SchemaEnsureResult:
    manifest: str
    fingerprint: str
    database: Optional[str]
    required_count: int
    verified_count: int
    elapsed_seconds: float


def _provider_name(driver: Any) -> str:
    provider = getattr(driver, "provider", "unknown")
    return str(getattr(provider, "value", provider)).lower()


def _normalize_properties(value: Any) -> Tuple[str, ...]:
    if isinstance(value, str):
        return (value,)
    if isinstance(value, Sequence):
        return tuple(str(item) for item in value)
    return ()


def _normalize_index_type(value: Any) -> str:
    index_type = str(value or "").casefold()
    if index_type == "btree":
        return "range"
    return index_type


def _normalize_entity_type(value: Any) -> str:
    return str(value or "node").casefold()


def _index_record(
    record: Mapping[str, Any],
) -> Tuple[str, Tuple[str, ...], str, str, str]:
    label_value = record.get("label", record.get("labelsOrTypes", record.get("labels_or_types", "")))
    if isinstance(label_value, Sequence) and not isinstance(label_value, str):
        label = str(label_value[0]) if label_value else ""
    else:
        label = str(label_value or "")
    properties = _normalize_properties(record.get("properties", record.get("property", ())))
    index_type = _normalize_index_type(record.get("index_type", record.get("type", "")))
    entity_type = _normalize_entity_type(
        record.get("entity_type", record.get("entityType", record.get("entitytype", "node")))
    )
    status = str(record.get("status", record.get("state", "")) or "").upper()
    return label, properties, index_type, entity_type, status


def _required_status(
    indexes: Iterable[SchemaIndex], records: Iterable[Mapping[str, Any]]
) -> Tuple[list[SchemaIndex], list[Tuple[SchemaIndex, str]]]:
    available: Dict[Tuple[str, Tuple[str, ...], str, str], str] = {}
    for record in records:
        label, properties, index_type, entity_type, status = _index_record(record)
        available[(label, properties, index_type, entity_type)] = status
    pending: list[SchemaIndex] = []
    failed: list[Tuple[SchemaIndex, str]] = []
    for index in indexes:
        if not index.required:
            continue
        status = available.get(index.key, "")
        if status in _FAILED_STATES:
            failed.append((index, status))
        elif status not in _READY_STATES:
            pending.append(index)
    return pending, failed


def _available_status(
    records: Iterable[Mapping[str, Any]],
) -> Dict[Tuple[str, Tuple[str, ...], str, str], str]:
    available: Dict[Tuple[str, Tuple[str, ...], str, str], str] = {}
    for record in records:
        label, properties, index_type, entity_type, status = _index_record(record)
        available[(label, properties, index_type, entity_type)] = status
    return available


async def ensure_schema(
    driver: Any,
    manifest: GraphSchemaManifest = CODE_GRAPH_SCHEMA,
    *,
    database: Optional[str] = None,
    timeout_seconds: float = 60.0,
    poll_interval_seconds: float = 0.1,
) -> SchemaEnsureResult:
    """Create and verify required indexes before the first graph mutation."""

    started = time.monotonic()
    cache_key = (_provider_name(driver), database, manifest.fingerprint)
    cache = getattr(driver, "_schema_preflight_cache", None)
    if cache is None:
        cache = {}
        setattr(driver, "_schema_preflight_cache", cache)
    cached = cache.get(cache_key)
    if cached is not None:
        return cached

    lock = getattr(driver, "_schema_preflight_lock", None)
    if lock is None:
        lock = asyncio.Lock()
        setattr(driver, "_schema_preflight_lock", lock)

    async with lock:
        cached = cache.get(cache_key)
        if cached is not None:
            return cached
        inspect_indexes = getattr(driver, "inspect_indexes", None)
        if not callable(inspect_indexes):
            raise SchemaPreflightError(
                f"schema {manifest.name}@{manifest.fingerprint} cannot be verified: "
                "driver does not implement inspect_indexes"
            )

        if _provider_name(driver) == "falkordb":
            unsupported = [index for index in manifest.indexes if len(index.properties) != 1]
            if unsupported:
                detail = ", ".join(
                    f"{index.label}({','.join(index.properties)})" for index in unsupported
                )
                raise SchemaPreflightError(
                    "FalkorDB schema manifests require single-property indexes: " + detail
                )

        deadline = started + timeout_seconds
        try:
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                raise TimeoutError
            initial_records = await asyncio.wait_for(
                inspect_indexes(database=database), timeout=remaining
            )
        except TimeoutError as exc:
            raise SchemaPreflightError(
                f"schema {manifest.name}@{manifest.fingerprint} inspection exceeded "
                f"{timeout_seconds:.1f}s deadline"
            ) from exc
        except Exception as exc:
            raise SchemaPreflightError(
                f"schema {manifest.name}@{manifest.fingerprint} inspection failed: {exc}"
            ) from exc
        available = _available_status(initial_records)
        failed_initial = [
            (index, available[index.key])
            for index in manifest.indexes
            if index.required and available.get(index.key) in _FAILED_STATES
        ]
        if failed_initial:
            detail = ", ".join(
                f"{index.label}({','.join(index.properties)})/{index.index_type}={status}"
                for index, status in failed_initial
            )
            raise SchemaPreflightError(
                f"schema {manifest.name}@{manifest.fingerprint} contains failed required indexes: "
                f"{detail}"
            )

        missing_required = [
            index for index in manifest.indexes if index.required and index.key not in available
        ]
        missing_optional = [
            index for index in manifest.indexes if not index.required and index.key not in available
        ]
        for index in missing_required:
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                raise SchemaPreflightError(
                    f"schema {manifest.name}@{manifest.fingerprint} creation exceeded "
                    f"{timeout_seconds:.1f}s deadline before {index.label}"
                )
            logger.info(
                "Graph schema index create_started database=%s label=%s properties=%s type=%s",
                database,
                index.label,
                ",".join(index.properties),
                index.index_type,
            )
            try:
                await asyncio.wait_for(
                    driver.create_indexes(manifest.driver_indexes((index,)), database=database),
                    timeout=remaining,
                )
            except TimeoutError as exc:
                raise SchemaPreflightError(
                    f"schema {manifest.name}@{manifest.fingerprint} creation timed out for "
                    f"{index.label}({','.join(index.properties)}); completion is ambiguous and "
                    "must be re-inspected before retry"
                ) from exc
            except Exception as exc:
                raise SchemaPreflightError(
                    f"schema {manifest.name}@{manifest.fingerprint} creation failed for "
                    f"{index.label}({','.join(index.properties)}): {exc}"
                ) from exc
            logger.info(
                "Graph schema index create_finished database=%s label=%s properties=%s type=%s",
                database,
                index.label,
                ",".join(index.properties),
                index.index_type,
            )

        for index in missing_optional:
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                logger.warning("Optional graph index skipped after schema deadline: %s", index)
                continue
            try:
                await asyncio.wait_for(
                    driver.create_indexes(manifest.driver_indexes((index,)), database=database),
                    timeout=remaining,
                )
            except Exception as exc:
                logger.warning("Optional graph index creation failed for %s: %s", index, exc)

        pending = [index for index in manifest.indexes if index.required]
        while pending:
            try:
                records = (
                    initial_records
                    if not missing_required and pending == [index for index in manifest.indexes if index.required]
                    else await asyncio.wait_for(
                        inspect_indexes(database=database),
                        timeout=max(0.001, deadline - time.monotonic()),
                    )
                )
                initial_records = []
            except TimeoutError as exc:
                raise SchemaPreflightError(
                    f"schema {manifest.name}@{manifest.fingerprint} inspection exceeded "
                    f"{timeout_seconds:.1f}s deadline"
                ) from exc
            except Exception as exc:
                raise SchemaPreflightError(
                    f"schema {manifest.name}@{manifest.fingerprint} inspection failed: {exc}"
                ) from exc
            pending, failed = _required_status(manifest.indexes, records)
            if failed:
                detail = ", ".join(
                    f"{index.label}({','.join(index.properties)})/{index.index_type}={status}"
                    for index, status in failed
                )
                raise SchemaPreflightError(
                    f"schema {manifest.name}@{manifest.fingerprint} contains failed required indexes: "
                    f"{detail}"
                )
            if not pending:
                break
            if time.monotonic() >= deadline:
                detail = ", ".join(
                    f"{index.label}({','.join(index.properties)})" for index in pending
                )
                raise SchemaPreflightError(
                    f"schema {manifest.name}@{manifest.fingerprint} not operational before "
                    f"{timeout_seconds:.1f}s deadline: {detail}"
                )
            await asyncio.sleep(poll_interval_seconds)

        result = SchemaEnsureResult(
            manifest=manifest.name,
            fingerprint=manifest.fingerprint,
            database=database,
            required_count=sum(index.required for index in manifest.indexes),
            verified_count=sum(index.required for index in manifest.indexes),
            elapsed_seconds=time.monotonic() - started,
        )
        cache[cache_key] = result
        logger.info(
            "Graph schema ready provider=%s database=%s manifest=%s fingerprint=%s indexes=%d elapsed=%.3fs",
            _provider_name(driver),
            database,
            manifest.name,
            manifest.fingerprint,
            result.verified_count,
            result.elapsed_seconds,
        )
        return result


__all__ = ["SchemaEnsureResult", "SchemaPreflightError", "ensure_schema"]

File: rust/crates/cortex-graph-writer/src/preflight.rs
//! Port của `tools/graph/schema/preflight.py` — automatic, fail-closed graph
//! schema preflight: inspect → create missing required → poll đến ONLINE.

use std::collections::BTreeMap;
use std::time::{Duration, Instant};

use serde_json::Value;

use cortex_graph_core::schema_manifest::{code_graph_schema, GraphSchemaManifest};

use crate::store::{GraphStore, IndexSpec, StoreError};

const READY_STATES: [&str; 3] = ["ONLINE", "OPERATIONAL", "READY"];
const FAILED_STATES: [&str; 3] = ["FAILED", "FAILURE", "ERROR"];

/// `SchemaEnsureResult`.
#[derive(Debug, Clone)]
pub struct SchemaEnsureResult {
    pub manifest: String,
    pub fingerprint: String,
    pub database: Option<String>,
    pub required_count: usize,
    pub verified_count: usize,
    pub elapsed_seconds: f64,
}

fn index_record(record: &BTreeMap<String, Value>) -> (String, Vec<String>, String, String, String) {
    let label_value = record
        .get("label")
        .or_else(|| record.get("labelsOrTypes"))
        .cloned()
        .unwrap_or(Value::Null);
    let label = match &label_value {
        Value::Array(items) => items
            .first()
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string(),
        Value::String(s) => s.clone(),
        _ => String::new(),
    };
    let properties_value = record
        .get("properties")
        .or_else(|| record.get("property"))
        .cloned()
        .unwrap_or(Value::Null);
    let properties: Vec<String> = match &properties_value {
        Value::String(s) => vec![s.clone()],
        Value::Array(items) => items
            .iter()
            .filter_map(Value::as_str)
            .map(str::to_string)
            .collect(),
        _ => vec![],
    };
    let index_type_raw = record
        .get("index_type")
        .or_else(|| record.get("type"))
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_lowercase();
    let index_type = if index_type_raw == "btree" {
        "range".to_string()
    } else {
        index_type_raw
    };
    let entity_type = record
        .get("entity_type")
        .or_else(|| record.get("entityType"))
        .or_else(|| record.get("entitytype"))
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .unwrap_or("node")
        .to_lowercase();
    let status = record
        .get("status")
        .or_else(|| record.get("state"))
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_uppercase();
    (label, properties, index_type, entity_type, status)
}

type IndexKey = (String, Vec<String>, String, String);

fn available_status(records: &[BTreeMap<String, Value>]) -> BTreeMap<IndexKey, String> {
    let mut available = BTreeMap::new();
    for record in records {
        let (label, properties, index_type, entity_type, status) = index_record(record);
        available.insert((label, properties, index_type, entity_type), status);
    }
    available
}

/// `ensure_schema` — create + verify required indexes trước mutation đầu tiên.
/// (Không có asyncio: poll đồng bộ với deadline.)
pub fn ensure_schema(
    store: &mut dyn GraphStore,
    database: Option<&str>,
) -> Result<SchemaEnsureResult, StoreError> {
    ensure_schema_with(store, &code_graph_schema(), database, Duration::from_secs(60))
}

/// Bản manifest-tường-minh của `ensure_schema` (unit-testable).
pub fn ensure_schema_with(
    store: &mut dyn GraphStore,
    manifest: &GraphSchemaManifest,
    database: Option<&str>,
    timeout: Duration,
) -> Result<SchemaEnsureResult, StoreError> {
    let started = Instant::now();
    let deadline = started + timeout;

    if store.provider() == "falkordb" {
        let unsupported: Vec<&cortex_graph_core::schema_manifest::SchemaIndex> = manifest
            .indexes
            .iter()
            .filter(|index| index.properties.len() != 1)
            .collect();
        if !unsupported.is_empty() {
            let detail = unsupported
                .iter()
                .map(|index| format!("{}({})", index.label, index.properties.join(",")))
                .collect::<Vec<_>>()
                .join(", ");
            return Err(StoreError::Schema(format!(
                "FalkorDB schema manifests require single-property indexes: {detail}"
            )));
        }
    }

    let initial_records = store.inspect_indexes(database)?;
    let available = available_status(&initial_records);

    // Required indexes đã FAILED từ đầu → fail ngay (không create lại).
    let failed_initial: Vec<String> = manifest
        .indexes
        .iter()
        .filter(|index| {
            index.required
                && available
                    .get(&(
                        index.label.clone(),
                        index.properties.clone(),
                        index.index_type.clone(),
                        index.entity_type.clone(),
                    ))
                    .map(|status| FAILED_STATES.contains(&status.as_str()))
                    .unwrap_or(false)
        })
        .map(|index| {
            format!(
                "{}({})/{}",
                index.label,
                index.properties.join(","),
                index.index_type
            )
        })
        .collect();
    if !failed_initial.is_empty() {
        return Err(StoreError::Schema(format!(
            "schema {}@{} contains failed required indexes: {}",
            manifest.name,
            manifest.fingerprint(),
            failed_initial.join(", ")
        )));
    }

    let key = |index: &cortex_graph_core::schema_manifest::SchemaIndex| -> IndexKey {
        (
            index.label.clone(),
            index.properties.clone(),
            index.index_type.clone(),
            index.entity_type.clone(),
        )
    };

    let missing_required: Vec<&cortex_graph_core::schema_manifest::SchemaIndex> = manifest
        .indexes
        .iter()
        .filter(|index| index.required && !available.contains_key(&key(index)))
        .collect();
    let missing_optional: Vec<&cortex_graph_core::schema_manifest::SchemaIndex> = manifest
        .indexes
        .iter()
        .filter(|index| !index.required && !available.contains_key(&key(index)))
        .collect();

    for index in &missing_required {
        if Instant::now() >= deadline {
            return Err(StoreError::Schema(format!(
                "schema {}@{} creation exceeded {:.1}s deadline before {}",
                manifest.name,
                manifest.fingerprint(),
                timeout.as_secs_f64(),
                index.label
            )));
        }
        store.create_indexes(
            &[IndexSpec {
                label: index.label.clone(),
                property: index.properties[0].clone(),
                index_type: index.index_type.clone(),
            }],
            database,
        )?;
    }

    for index in &missing_optional {
        if Instant::now() >= deadline {
            break;
        }
        let _ = store.create_indexes(
            &[IndexSpec {
                label: index.label.clone(),
                property: index.properties[0].clone(),
                index_type: index.index_type.clone(),
            }],
            database,
        );
    }

    // Poll đến khi mọi required index ONLINE.
    let mut inspected = initial_records;
    if !missing_required.is_empty() {
        inspected = store.inspect_indexes(database)?;
    }
    loop {
        let available = available_status(&inspected);
        let mut pending: Vec<String> = Vec::new();
        let mut failed: Vec<String> = Vec::new();
        for index in manifest.indexes.iter().filter(|index| index.required) {
            let status = available
                .get(&key(index))
                .map(String::as_str)
                .unwrap_or("");
            if FAILED_STATES.contains(&status) {
                failed.push(format!(
                    "{}({})/{}={status}",
                    index.label,
                    index.properties.join(","),
                    index.index_type
                ));
            } else if !READY_STATES.contains(&status) {
                pending.push(format!(
                    "{}({})",
                    index.label,
                    index.properties.join(",")
                ));
            }
        }
        if !failed.is_empty() {
            return Err(StoreError::Schema(format!(
                "schema {}@{} contains failed required indexes: {}",
                manifest.name,
                manifest.fingerprint(),
                failed.join(", ")
            )));
        }
        if pending.is_empty() {
            break;
        }
        if Instant::now() >= deadline {
            return Err(StoreError::Schema(format!(
                "schema {}@{} not operational before {:.1}s deadline: {}",
                manifest.name,
                manifest.fingerprint(),
                timeout.as_secs_f64(),
                pending.join(", ")
            )));
        }
        std::thread::sleep(Duration::from_millis(100));
        inspected = store.inspect_indexes(database)?;
    }

    let required_count = manifest.indexes.iter().filter(|index| index.required).count();
    Ok(SchemaEnsureResult {
        manifest: manifest.name.clone(),
        fingerprint: manifest.fingerprint(),
        database: database.map(str::to_string),
        required_count,
        verified_count: required_count,
        elapsed_seconds: started.elapsed().as_secs_f64(),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn index_record_normalizes_labels_and_types() {
        let mut record = BTreeMap::new();
        record.insert("label".to_string(), Value::String("Function".into()));
        record.insert(
            "properties".to_string(),
            Value::Array(vec![Value::String("id".into())]),
        );
        record.insert("index_type".to_string(), Value::String("btree".into()));
        record.insert("status".to_string(), Value::String("online".into()));
        let (label, properties, index_type, entity_type, status) = index_record(&record);
        assert_eq!(label, "Function");
        assert_eq!(properties, vec!["id".to_string()]);
        assert_eq!(index_type, "range");
        assert_eq!(entity_type, "node");
        assert_eq!(status, "ONLINE");
    }

}
